src/lib/faunadb.mjs

```diff
@@ -0,0 +1,158 @@
+// @ts-check
+import { createContext, runInContext } from "node:vm";
+
+import { container } from "../cli.mjs";
+import { formatFullErrorForShell, formatObjectForShell } from "./misc.mjs";
+
+/**
+ * Creates a V4 Fauna client.
+ * @param {object} argv
+ * @param {string} argv.url
+ * @param {string} argv.secret
+ * @param {number} [argv.timeout]
+ * @returns {Promise<import("faunadb").Client>}
+ */
+export const getClient = async (argv) => {
+  const { Client } = container.resolve("faunadb");
+  const { hostname, port, protocol } = new URL(argv.url);
+  const scheme = protocol?.replace(/:$/, "");
+  
+  return new Client({
+    domain: hostname,
+    port: Number(port),
+    scheme: /** @type {('http'|'https')} */ (scheme),
+    secret: argv.secret,
+    timeout: argv.timeout,
+    fetch: container.resolve("fetch"),
+    headers: {
+      "x-fauna-shell-builtin": "true",
+      "x-fauna-source": "Fauna Shell",
+    }
+  });
+};
+
+/**
+ * Interprets a string as a v4 FQL expression and returns a query.
+ * @param {string} expression - The FQL expression to interpret
+ * @returns {Promise<import("faunadb").Expr>} The resulting query
+ */
+export async function stringExpressionToQuery(expression) {
+  const faunadb = (await import("faunadb")).default;
+  return runInContext(expression, createContext(faunadb.query));
+}
+
+const validateQueryParams = ({ query, client, url, secret }) => {
+  if (!query) {
+    throw new Error("A query is required.");
+  } else if (!client && (!url || !secret)) {
+    throw new Error("A client or url and secret are required.");
+  }
+};
```

Comment:
Saw some validation of args above somewhere as well. I have some arg validation in Credentials class too. 

Thoughts on validating all args and their combinations in one place in middleware before we get to handlers?

Reply:
In theory, this check doesn't need to be here at all. The fauna client will barf if we're missing things. This error isn't adding a lot of value as-is.